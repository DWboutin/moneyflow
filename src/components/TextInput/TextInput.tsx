import React from "react";
import styled from "styled-components";
import { ThemeContainer } from "../../styles/styles";

interface ContainerProps {}

export const Container = styled.div<ContainerProps>`
  display: flex;
  flex-direction: column;
  flex: 1;

  label {
    margin-bottom: 10px;
  }
`;

const Input = styled.input`
  height: 40px;
  padding: 10px 20px 9px 20px;
  border-radius: 0px 20px 20px 20px;
  color: ${({ theme }: ThemeContainer) => theme.input.text};
  border: 1px solid ${({ theme }: ThemeContainer) => theme.input.normal.border};
  background-color: ${({ theme }: ThemeContainer) => theme.input.normal.bg};

  &::placeholder {
    color: ${({ theme }: ThemeContainer) => theme.input.placeholder};
  }

  &:focus-visible {
    border: 1px solid ${({ theme }: ThemeContainer) => theme.input.focus.border};
    outline: none;
  }
`;

const Error = styled.div`
  padding-top: 5px;
  color: ${({ theme }: ThemeContainer) => theme.input.error};
`;

export interface Props {
  id: string;
  name: string;
  label: string;
  placeholder: string;
  value: string | number;
  handleOnChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  error?: string | boolean;
}

const TextInput: React.VoidFunctionComponent<Props> = ({
  id,
  name,
  label,
  placeholder,
  value,
  handleOnChange,
  error,
}) => {
  return (
    <Container>
      <label htmlFor={id}>{label}</label>
      <Input
        type="text"
        id={id}
        name={name}
        placeholder={placeholder}
        value={value}
        onChange={handleOnChange}
      />
      {error && <Error>{error}</Error>}
    </Container>
  );
};

export default TextInput;
