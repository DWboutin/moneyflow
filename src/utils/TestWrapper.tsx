import React, { ReactNode } from "react";
import { Provider } from "react-redux";
import { ToastContainer } from "react-toastify";
import { ThemeProvider } from "styled-components";

import store from "../store/store";
import GlobalStyle from "../styles/globalStyle";

import { theme } from "../styles/styles";

interface Props {
  children: ReactNode;
}

const TestWrapper: React.FunctionComponent<Props> = ({ children }) => (
  <Provider store={store}>
    <ThemeProvider theme={theme}>
      <GlobalStyle />
      {children}
      <ToastContainer />
    </ThemeProvider>
  </Provider>
);

export default TestWrapper;
