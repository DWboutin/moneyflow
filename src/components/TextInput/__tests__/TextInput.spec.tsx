import React from "react";
import { fireEvent, render, RenderResult } from "@testing-library/react";

import TextInput, { Props } from "../TextInput";
import TestWrapper from "../../../utils/TestWrapper";

describe("<TextInput />", () => {
  let renderResult: RenderResult;

  const mockOnChange = jest.fn();
  const renderTextInput = ({
    id,
    name,
    label,
    placeholder,
    value,
    handleOnChange,
    error,
  }: Props) => {
    renderResult = render(
      <TestWrapper>
        <TextInput
          id={id}
          name={name}
          label={label}
          placeholder={placeholder}
          value={value}
          handleOnChange={handleOnChange}
          error={error}
        />
      </TestWrapper>
    );
  };

  describe("default rendering", () => {
    beforeEach(() => {
      renderTextInput({
        id: "id",
        name: "name",
        label: "Label",
        placeholder: "Placeholder",
        value: "",
        handleOnChange: mockOnChange,
      });
    });

    it("should match snapshot", () => {
      expect(renderResult.container).toMatchSnapshot();
    });
  });

  describe("with error", () => {
    beforeEach(() => {
      renderTextInput({
        id: "id",
        name: "name",
        label: "Label",
        placeholder: "Placeholder",
        value: "Value",
        handleOnChange: mockOnChange,
        error: "Has an error",
      });
    });

    it("should match snapshot", () => {
      expect(renderResult.container).toMatchSnapshot();
    });
  });

  describe("on change", () => {
    beforeEach(() => {
      renderTextInput({
        id: "id",
        name: "name",
        label: "Label",
        placeholder: "Placeholder",
        value: "",
        handleOnChange: mockOnChange,
      });

      fireEvent.change(
        renderResult.container.querySelector("input") as HTMLTextAreaElement,
        {
          target: { value: "New value" },
        }
      );
    });

    it("should execute handleOnChange", async () => {
      expect(mockOnChange).toHaveBeenCalled();
    });
  });
});
