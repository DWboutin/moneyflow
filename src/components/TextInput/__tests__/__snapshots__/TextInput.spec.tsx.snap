// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextInput /> default rendering should match snapshot 1`] = `
<div>
  <div
    class="sc-bdvvtL kGCSBy"
  >
    <label
      for="id"
    >
      Label
    </label>
    <input
      class="sc-gsDKAQ QPTvR"
      id="id"
      name="name"
      placeholder="Placeholder"
      type="text"
      value=""
    />
  </div>
  <div
    class="Toastify"
  />
</div>
`;

exports[`<TextInput /> with error should match snapshot 1`] = `
<div>
  <div
    class="sc-bdvvtL kGCSBy"
  >
    <label
      for="id"
    >
      Label
    </label>
    <input
      class="sc-gsDKAQ QPTvR"
      id="id"
      name="name"
      placeholder="Placeholder"
      type="text"
      value="Value"
    />
    <div
      class="sc-dkPtRN AKYuc"
    >
      Has an error
    </div>
  </div>
  <div
    class="Toastify"
  />
</div>
`;
